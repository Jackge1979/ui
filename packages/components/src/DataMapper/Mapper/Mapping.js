import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Mapping extends Component {
	constructor(props) {
		super(props);
		this.state = {
			width: 1,
			height: 1,
		};
		this.updateMappingAreaRef = this.updateMappingAreaRef.bind(this);
		this.updateMappingContentRef = this.updateMappingContentRef.bind(this);
		this.checkMappingContentResize = this.checkMappingContentResize.bind(this);
	}

	componentDidMount() {
		this.intervalRef = setInterval(this.checkMappingContentResize, 250);
	}

	componentWillUnmount() {
		clearInterval(this.intervalRef);
	}

	getArea() {
		if (this.mappingAreaRef) {
			let area = this.mappingAreaRef;
			if (this.mappingAreaRef.getDecoratedComponentInstance) {
				area = this.mappingAreaRef.getDecoratedComponentInstance();
			}
			return area;
		}
		return {
			update() {
				// LOG
			},
			reveal() {
				// LOG
			},
			getMousePos(offset) {
				return offset;
			},
		};
	}

	checkMappingContentResize() {
		if (this.mappingContentRef) {
			const width = this.mappingContentRef.clientWidth;
			const height = this.mappingContentRef.clientHeight;
			if (this.state.width !== width || this.state.height !== height) {
				this.setState({ width, height });
			}
		}
	}

	update() {
		this.getArea().update();
	}

	reveal(connectionKey) {
		this.getArea().reveal(connectionKey);
	}

	beginDrag(element, side) {
		return this.props.dndListener.beginDrag(element, side);
	}

	dndInProgress(offset) {
		const pos = this.getArea().getMousePos(offset);
		this.props.dndListener.dndInProgress(pos);
	}

	canDrop(sourceItem, targetItem) {
		return this.props.dndListener.canDrop(sourceItem, targetItem);
	}

	drop(sourceItem, targetItem) {
		this.props.dndListener.drop(sourceItem, targetItem);
	}

	endDrag() {
		this.props.dndListener.endDrag();
	}

	updateMappingAreaRef(ref) {
		this.mappingAreaRef = ref;
	}

	updateMappingContentRef(ref) {
		this.mappingContentRef = ref;
	}

	render() {
		const { dndListener, ...mappingProps } = this.props;
		const { mappingConfiguration } = this.props;
		const MappingActions = mappingConfiguration.getActions();
		const MappingRenderer = mappingConfiguration.getRenderer();
		return (
			<div className="mapping mapper-element">
				<MappingActions {...this.props} />
				<div className="separator horizontal" />
				<div ref={this.updateMappingContentRef} className="mapping-content">
					<MappingRenderer
						{...mappingProps}
						ref={this.updateMappingAreaRef}
						dndListener={this}
						width={this.state.width}
						height={this.state.height}
					/>
				</div>
			</div>
		);
	}
}

Mapping.propTypes = {
	mappingConfiguration: PropTypes.object,
	dndListener: PropTypes.object,
};
