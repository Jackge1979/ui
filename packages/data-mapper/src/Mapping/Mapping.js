import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Actions } from '@talend/react-components';
import MappingSVG from './MappingSVG';

export default class Mapping extends Component {
	constructor(props) {
		super(props);
		this.state = {
			width: 1,
			height: 1,
		};
		this.updateMappingAreaRef = this.updateMappingAreaRef.bind(this);
		this.updateMappingContentRef = this.updateMappingContentRef.bind(this);
		this.checkMappingContentResize = this.checkMappingContentResize.bind(this);
	}

	componentDidMount() {
		this.intervalRef = setInterval(this.checkMappingContentResize, 250);
	}

	componentWillUnmount() {
		clearInterval(this.intervalRef);
	}

	getArea() {
		if (this.mappingAreaRef) {
			let area = this.mappingAreaRef;
			if (this.mappingAreaRef.getDecoratedComponentInstance) {
				area = this.mappingAreaRef.getDecoratedComponentInstance();
			}
			return area;
		}
		return {
			update() {
				// LOG
			},
			reveal() {
				// LOG
			},
			getMousePos(offset) {
				return offset;
			},
		};
	}

	checkMappingContentResize() {
		if (this.mappingContentRef) {
			const width = this.mappingContentRef.clientWidth;
			const height = this.mappingContentRef.clientHeight;
			if (this.state.width !== width || this.state.height !== height) {
				this.setState({ width, height });
			}
		}
	}

	update() {
		this.getArea().update();
	}

	updateDND() {
		if (this.getArea().updateDND) {
			this.getArea().updateDND();
		}
	}

	reveal(connectionKey) {
		this.getArea().reveal(connectionKey);
	}

	beginDrag(element, side) {
		return this.props.dndListener.beginDrag(element, side);
	}

	dndInProgress(offset) {
		const pos = this.getArea().getMousePos(offset);
		this.props.dndListener.dndInProgress(pos);
	}

	canDrop(sourceItem, targetItem) {
		return this.props.dndListener.canDrop(sourceItem, targetItem);
	}

	drop(sourceItem, targetItem) {
		this.props.dndListener.drop(sourceItem, targetItem);
	}

	endDrag() {
		this.props.dndListener.endDrag();
	}

	updateMappingAreaRef(ref) {
		this.mappingAreaRef = ref;
	}

	updateMappingContentRef(ref) {
		this.mappingContentRef = ref;
	}

	render() {
		const { dndListener, mappingActions, ...mappingProps } = this.props;
		return (
			<div className="mapping mapper-element">
				<Actions className="mapping-actions" actions={mappingActions} />
				<div ref={this.updateMappingContentRef} className="mapping-content">
					<MappingSVG
						{...mappingProps}
						ref={this.updateMappingAreaRef}
						dndListener={this}
						width={this.state.width}
						height={this.state.height}
					/>
				</div>
			</div>
		);
	}
}

Mapping.propTypes = {
	mappingActions: PropTypes.object,
	dndListener: PropTypes.object,
};
