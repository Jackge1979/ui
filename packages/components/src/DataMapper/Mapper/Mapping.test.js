import React from 'react';
import Mapping from './Mapping.js'
import renderer from 'react-test-renderer';

const mapping = [
  {
    source: 'elem_in_1',
    target: 'elem_out_1'
  },
  {
    source: 'elem_in_2',
    target: 'elem_out_2'
  },
  {
    source: 'elem_in_3',
    target: 'elem_out_3'
  }
]

it('mapping', () => {
  // create React tree
  const tree = renderer
    .create(<Mapping mapping={mapping} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
