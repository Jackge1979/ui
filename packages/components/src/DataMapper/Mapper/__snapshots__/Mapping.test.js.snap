// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no connections - no anchors 1`] = `
<div
  className="mapping mapper-element"
>
  <div
    className="tc-actions mapping-tools btn-group"
  >
    <button
      aria-label="Show"
      className="btn btn-default"
      disabled={false}
      id="show-all"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Show
      </span>
    </button>
    <button
      aria-label="Clear"
      className="btn btn-default"
      disabled={false}
      id="clear-connection"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear
      </span>
    </button>
    <button
      aria-label="Clear All"
      className="btn btn-default"
      disabled={false}
      id="clear-mapping"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear All
      </span>
    </button>
  </div>
  <div
    className="separator horizontal"
  />
  <div
    className="mapping-content"
  >
    <svg
      className="mapping-svg"
      height={1}
      width={26}
      xmlns="http://www.w3.org/2000/svg"
    />
  </div>
</div>
`;

exports[`no connections - unmapped anchors 1`] = `
<div
  className="mapping mapper-element"
>
  <div
    className="tc-actions mapping-tools btn-group"
  >
    <button
      aria-label="Show"
      className="btn btn-default"
      disabled={false}
      id="show-all"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Show
      </span>
    </button>
    <button
      aria-label="Clear"
      className="btn btn-default"
      disabled={false}
      id="clear-connection"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear
      </span>
    </button>
    <button
      aria-label="Clear All"
      className="btn btn-default"
      disabled={false}
      id="clear-mapping"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear All
      </span>
    </button>
  </div>
  <div
    className="separator horizontal"
  />
  <div
    className="mapping-content"
  >
    <svg
      className="mapping-svg"
      height={1}
      width={26}
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="anchor start unmapped"
        cx={8}
        cy={15}
        r={4}
      />
      <circle
        className="anchor start unmapped"
        cx={8}
        cy={40}
        r={4}
      />
      <circle
        className="anchor end unmapped"
        cx={18}
        cy={15}
        r={4}
      />
      <circle
        className="anchor end unmapped"
        cx={18}
        cy={40}
        r={4}
      />
    </svg>
  </div>
</div>
`;

exports[`single connection [all] - mapped/unmapped anchors 1`] = `
<div
  className="mapping mapper-element"
>
  <div
    className="tc-actions mapping-tools btn-group"
  >
    <button
      aria-label="Show"
      className="btn btn-default"
      disabled={false}
      id="show-all"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Show
      </span>
    </button>
    <button
      aria-label="Clear"
      className="btn btn-default"
      disabled={false}
      id="clear-connection"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear
      </span>
    </button>
    <button
      aria-label="Clear All"
      className="btn btn-default"
      disabled={false}
      id="clear-mapping"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear All
      </span>
    </button>
  </div>
  <div
    className="separator horizontal"
  />
  <div
    className="mapping-content"
  >
    <svg
      className="mapping-svg"
      height={1}
      width={26}
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="anchor start unmapped"
        cx={8}
        cy={40}
        r={4}
      />
      <circle
        className="anchor end unmapped"
        cx={18}
        cy={15}
        r={4}
      />
      <circle
        className="anchor start mapped"
        cx={8}
        cy={15}
        r={4}
      />
      <path
        className="anchor end mapped"
        d="M 13 35 L 23 40 L 13 45 Z"
      />
      <g
        id="in_1-out_2-all"
      >
        <circle
          className="connection start all"
          cx={8}
          cy={15}
          r={3}
        />
        <path
          className="connection curve all "
          d="M 8 15 C 13 15, 13 40, 18 40"
        >
          <animate
            attributeName="stroke-width"
            attributeType="CSS"
            begin="indefinite"
            dur="0.2s"
            from="3"
            id="anim-in_1-out_2-all"
            to="7"
          />
          <animate
            attributeName="stroke-width"
            attributeType="CSS"
            begin="anim-in_1-out_2-all.begin + 0.2s"
            dur="0.2s"
            from="7"
            to="3"
          />
        </path>
        <path
          className="connection end all"
          d="M 13 35 L 23 40 L 13 45 Z"
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`single connection [all] - no anchors 1`] = `
<div
  className="mapping mapper-element"
>
  <div
    className="tc-actions mapping-tools btn-group"
  >
    <button
      aria-label="Show"
      className="btn btn-default"
      disabled={false}
      id="show-all"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Show
      </span>
    </button>
    <button
      aria-label="Clear"
      className="btn btn-default"
      disabled={false}
      id="clear-connection"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear
      </span>
    </button>
    <button
      aria-label="Clear All"
      className="btn btn-default"
      disabled={false}
      id="clear-mapping"
      onClick={[Function]}
      onMouseDown={[Function]}
      role={null}
      type="button"
    >
      <span>
        Clear All
      </span>
    </button>
  </div>
  <div
    className="separator horizontal"
  />
  <div
    className="mapping-content"
  >
    <svg
      className="mapping-svg"
      height={1}
      width={26}
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        id="in_1-out_2-all"
      >
        <circle
          className="connection start all"
          cx={8}
          cy={15}
          r={3}
        />
        <path
          className="connection curve all "
          d="M 8 15 C 13 15, 13 40, 18 40"
        >
          <animate
            attributeName="stroke-width"
            attributeType="CSS"
            begin="indefinite"
            dur="0.2s"
            from="3"
            id="anim-in_1-out_2-all"
            to="7"
          />
          <animate
            attributeName="stroke-width"
            attributeType="CSS"
            begin="anim-in_1-out_2-all.begin + 0.2s"
            dur="0.2s"
            from="7"
            to="3"
          />
        </path>
        <path
          className="connection end all"
          d="M 13 35 L 23 40 L 13 45 Z"
        />
      </g>
    </svg>
  </div>
</div>
`;
